rest.c

```diff
@@ -142,6 +151,118 @@ restconf_error (int status, rest_e_tag error_tag)
     return output;
 }
 
+static void
+restconf_log_get_head (int flags, const char *path, const char *remote_user,
+                       const char *remote_addr, int rc)
+{
+    if ((flags & FLAGS_METHOD_GET) && restconf_logging_test_flag (LOG_GET))
+    {
+        NOTICE ("GET: %s@%s %s code:%u\n", remote_user, remote_addr, path, rc);
+    }
+    else if ((flags & FLAGS_METHOD_HEAD) && restconf_logging_test_flag (LOG_HEAD))
+    {
+        NOTICE ("HEAD: %s@%s %s code:%u\n", remote_user, remote_addr, path, rc);
+    }
+}
+
+static gboolean
+log_modified_leafs (GNode *node, gpointer arg)
+{
+    log_params *params = (log_params *) arg;
+    char *path = NULL;
+
+    path = apteryx_node_path (node);
+    if (node->data && strlen (node->data) != 0)
+        NOTICE ("%s: %s@%s %s%s code:%u\n",
+                params->key, params->remote_user, params->remote_addr, params->root_path, path, params->rc);
+
+    g_free (path);
+    return FALSE;
+}
+
+static gboolean
+log_deleted_leafs (GNode *node, gpointer arg)
+{
+    log_params *params = (log_params *) arg;
+    char *path = NULL;
+
+    if (!node->data || strlen (node->data) == 0)
+        node = node->parent;
+
+    if (node)
+    {
+        path = apteryx_node_path (node);
+        if (node->data)
+            NOTICE ("%s: %s@%s %s code:%u\n",
+                    params->key, params->remote_user, params->remote_addr, path, params->rc);
+    }
+    g_free (path);
+    return FALSE;
+}
+
+static void
+restconf_log_post_put_patch (int flags, const char *path, GNode *root, GNode *tree, const char *remote_user,
+                             const char *remote_addr, int rc)
+{
+    char *key = NULL;
+    if ((flags & FLAGS_METHOD_POST) && restconf_logging_test_flag (LOG_POST))
+        key = "POST";
+    else if ((flags & FLAGS_METHOD_PUT) && restconf_logging_test_flag (LOG_PUT))
+        key = "PUT";
+    else if ((flags & FLAGS_METHOD_PATCH) && restconf_logging_test_flag (LOG_PATCH))
+        key = "PATCH";
+
+    if (key)
+    {
+        log_params params;
+        GNode *rnode = root;
+
+        while (rnode && rnode->children)
+            rnode = rnode->children;
+
+        params.root_path = apteryx_node_path (rnode);
+        params.remote_user = remote_user;
+        params.remote_addr = remote_addr;
+        params.key = key;
+        params.rc = rc;
+
+        if (tree)
```

Comment:
why not just log the tree output ... this could generate thousands of log entries

Reply:
A proper audit log requires each individual change to the system configuration to be logged.

Reply:
what would you miss with the logging of the tree output. In fact the method above is more misleading as it suggests that there were individual set operations. Even from the CLI a single atomic change in configuration would often involve a tree set.